les/pruner.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package les
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// errPrunerClosed is returned if pruner is closed.
+var errPrunerClosed = errors.New("pruner is closed")
+
+// pruner is responsible for pruning historical light chain data.
+type pruner struct {
+	checkpoint uint64
+	closeCh    chan struct{}
+	addCh      chan *core.ChainIndexer
+	wg         sync.WaitGroup
+}
+
+// newPruner returns a light chain pruner instance.
+func newPruner(checkpoint uint64) *pruner {
+	pruner := &pruner{
+		checkpoint: checkpoint,
+		closeCh:    make(chan struct{}),
+		addCh:      make(chan *core.ChainIndexer),
+	}
+	pruner.wg.Add(1)
+	go pruner.loop()
+	return pruner
+}
+
+// register adds a batch of specific chain indexers as pruning target.
+func (p *pruner) register(indexers ...*core.ChainIndexer) error {
+	for _, indexer := range indexers {
+		select {
+		case p.addCh <- indexer:
+		case <-p.closeCh:
+			return errPrunerClosed
+		}
+	}
+	return nil
+}
+
+// close notifies all background goroutines belonging to pruner to exit.
+func (p *pruner) close() {
+	close(p.closeCh)
+	p.wg.Wait()
+}
+
+// loop periodically queries the status of chain indexers and prunes useless
+// historical chain data.
+func (p *pruner) loop() {
+	defer p.wg.Done()
+
+	var (
+		last     = p.checkpoint
+		ticker   = time.NewTicker(5 * time.Second)
+		indexers []*core.ChainIndexer
+	)
+	// pruning finds the sections that have been processed by all indexers
+	// and deletes all historical chain data.
+	// Note, if some indexers don't support pruning(e.g. eth.BloomIndexer),
+	// pruning operations can be silently ignored.
+	pruning := func() {
+		min := uint64(math.MaxUint64)
+		for _, indexer := range indexers {
+			sections, _, _ := indexer.Sections()
+			if sections < min {
+				min = sections
+			}
+		}
+		if min == 0 {
+			return
+		}
+		// Always keep the latest section data in database.
+		for i := last + 1; i < min-1; i++ {
```

Comment:
This logic does not handle indexer rollbacks. While this is not a frequent case we do handle it in the indexers (might happen on testnets/private nets). Rolling back while pruning is a bit tricky though (you would have to restore the headers of the current unfinished section) so I think it is fine to not be able to roll back properly if pruning is enabled. In this case an error message would be nice at least. And/or automatically reverting to the last stable checkpoint and resyncing from there.

Reply:
Now I'll keep the latest section in db. It means at least 32768+2048 headers are kept. 
Do we really have that deep reorg even in the testnet? Maybe it's possible in the ropsten...

Now we don't have a very good solution to restore all pruned chain data. Rewind HEAD to current reverted head seems feasible. I may try this approach.